File: python/snake_game/highscore.py
"""Persists the all-time high score to a small JSON file on disk."""
import json
import os


class HighScoreStore:
    def __init__(self, path):
        self.path = path
        self.value = self._load()

    def _load(self):
        if not os.path.exists(self.path):
            return 0
        try:
            with open(self.path, 'r') as f:
                return int(json.load(f).get('high_score', 0))
        except (json.JSONDecodeError, ValueError, OSError):
            return 0

    def update(self, score):
        if score <= self.value:
            return False
        self.value = score
        self._save()
        return True

    def _save(self):
        try:
            with open(self.path, 'w') as f:
                json.dump({'high_score': self.value}, f)
        except OSError:
            pass

File: cpp/src/highscore.cpp
#include "highscore.hpp"

#include <fstream>

#include "json.hpp"

HighScoreStore::HighScoreStore(std::string path) : path_(std::move(path)), value_(Load()) {}

int HighScoreStore::Load() const {
    std::ifstream in(path_);
    if (!in) return 0;
    try {
        nlohmann::json j;
        in >> j;
        return j.value("high_score", 0);
    } catch (const nlohmann::json::exception&) {
        return 0;
    }
}

bool HighScoreStore::Update(int score) {
    if (score <= value_) return false;
    value_ = score;
    Save();
    return true;
}

void HighScoreStore::Save() const {
    std::ofstream out(path_);
    if (!out) return;
    nlohmann::json j;
    j["high_score"] = value_;
    out << j.dump();
}
